src/target/riscv/riscv.c

```diff
@@ -1614,6 +1614,357 @@ static int riscv_hit_trigger_hit_bit(struct target *target, uint32_t *unique_id)
+	uint16_t uimm_5to4and8_bits = get_field32(instruction, INSN_FIELD_C_UIMM9HI);
+	uint16_t uimm_5to4_bits = uimm_5to4and8_bits >> 1;
+	uint16_t uimm_8_bit = uimm_5to4and8_bits & 0x1;
+	return (uimm_5to4_bits << 4) + (uimm_7to6_bits << 6) + (uimm_8_bit << 8);
+}
+
+// c.lqsp rd_n0 c_uimm10sphi c_uimm10splo - offset[5] offset[4|9:6]
+static inline uint16_t get_imm_clqsp(riscv_insn_t instruction) {
+	uint16_t uimm_4and9to6_bits =
+		get_field32(instruction, INSN_FIELD_C_UIMM10SPLO);
+	uint16_t uimm_4_bit = uimm_4and9to6_bits >> 4;
+	uint16_t uimm_9to6_bits = uimm_4and9to6_bits & 0xf;
+	uint16_t uimm_5_bit = get_field32(instruction, INSN_FIELD_C_UIMM10SPHI);
+	return (uimm_4_bit << 4) + (uimm_5_bit << 5) + (uimm_9to6_bits << 6);
+}
+
+// c.sqsp c_rs2 c_uimm10sp_s - offset[5:4|9:6]
+static inline uint16_t get_imm_csqsp(riscv_insn_t instruction) {
+	uint16_t uimm_5to4and9to6_bits =
+		get_field32(instruction, INSN_FIELD_C_UIMM10SP_S);
+	uint16_t uimm_5to4_biits = uimm_5to4and9to6_bits >> 4;
+	uint16_t uimm_9to6_bits = uimm_5to4and9to6_bits & 0xf;
+	return (uimm_5to4_biits << 4) + (uimm_9to6_bits << 6);
+}
+
+static inline uint32_t get_rs1_c(riscv_insn_t instruction) {
+	return 8 + get_field32(instruction, INSN_FIELD_C_SREG1);
+}
+
+static inline int get_memaddr_storeload(struct target *target,
+		const riscv_insn_t instruction, target_addr_t *mem_addr_wp) {
+	/* find out which memory address is accessed by the instruction at dpc */
+	/* opcode is first 7 bits of the instruction */
+	uint32_t opcode = instruction & INSN_FIELD_OPCODE;
+	if ((instruction & 0x03) < 0x03) { // opcode size RVC
+		/* RVC MASK_C = 0xe003 for load/store instructions */
+		opcode = instruction & MASK_C_LD;
+	}
+	uint32_t rs1;
+	int16_t imm;
+	target_addr_t mem_addr = 0;
+	bool misa_f = riscv_supports_extension(target, 'F');
+	bool misa_d = riscv_supports_extension(target, 'D');
+
+	switch (opcode) {
+	case MATCH_LB:
+		/* fallthrough */
+	case MATCH_FLH & ~INSN_FIELD_FUNCT3:
+		/* fallthrough */
+	case MATCH_SB:
+		/* fallthrough */
+	case MATCH_FSH & ~INSN_FIELD_FUNCT3:
+		rs1 = get_field32(instruction, INSN_FIELD_RS1);
+		if (riscv_get_register(target, &mem_addr, rs1) != ERROR_OK)
+			return ERROR_FAIL;
+		if (opcode == MATCH_SB || opcode == (MATCH_FSH & ~INSN_FIELD_FUNCT3)) {
+			LOG_TARGET_DEBUG(target, "%x is store instruction", instruction);
+			imm = get_field32(instruction, INSN_FIELD_IMM12LO) |
+				  (get_field32(instruction, INSN_FIELD_IMM12HI) << 5);
+		} else if (opcode == MATCH_LB ||
+				   opcode == (MATCH_FLH & ~INSN_FIELD_FUNCT3)) {
+			LOG_TARGET_DEBUG(target, "%x is load instruction", instruction);
+			imm = get_field32(instruction, INSN_FIELD_IMM12);
+		} else {
+			return ERROR_FAIL;
+		}
+		/* sign extend 12-bit imm to 16-bits */
+		if (imm & (1 << 11))
+			imm |= 0xf000;
+		mem_addr += imm;
+		LOG_TARGET_DEBUG(target, "Memory address=0x%" PRIx64, mem_addr);
+		break;
+
+	case MATCH_C_LWSP:
```

Comment:
[Suggestion]
AFAIU, in each case in this switch the following is done:
1. Number of the register, which contains the base of memory address is deduced from `instruction` and opcode.
2. The register is read.
3. Memory offset is deduced from the `instruction`
4. Memory address is base plus offset.

So, please, consider moving the logic/communication with target out of the switch.
What I mean is something along the lines of:
```
unsigned int rs = get_storeload_membase_regno(instruction);  /* Inside the function there is a `switch (opcode)`*/
if (riscv_get_register(target, &mem_addr, rs) != ERROR_OK)
    return ERROR_FAIL;
target_addr_t offset = get_storeload_memoffset(instruction); /* Inside the function there is a `switch (opcode)`*/
if (verify_storload(target, opcode, rs, offset) != ERROR_OK) /* This function checks that the deuced values are coherent (like 'the code points with rd=x0 are reserved' for MATCH_C_LDSP)*/
    return ERROR_FAIL;
```

IMHO, this will make the code a bit easier to read.